lib/format.c

```diff
@@ -0,0 +1,470 @@
+			break;
+		case TVAR:
+			r = parse(&pos, &buf, &spec);
+			if (r < 0)
+				goto err;
+			fmt[n].var = calloc(1, sizeof(struct var)+buf.len+1);
+			if (!fmt[n].var)
+				goto err_errno;
+			fmt[n].common->type = TVAR;
+			fmt[n].var->spec = spec;
+			memcpy(fmt[n].var->s, buf.mem, buf.len+1);
+			break;
+		}
+		n++;
+	}
+out:
+	strbuf_release(&buf);
+	return fmt;
+err_errno:
+	r = -errno;
+err:
+	free(fmt);
+	strbuf_release(&buf);
+	errno = -r;
+	return NULL;
+}
+
+void
+xbps_fmt_free(struct xbps_fmt *fmt)
+{
+	if (!fmt)
+		return;
+	for (struct xbps_fmt *f = fmt; f->common; f++)
+		switch (f->common->type) {
+		case TTEXT: free(f->chunk); break;
+		case TVAR: free(f->var); break;
+		}
+	free(fmt);
+}
+
+int
+xbps_fmts(const char *format, xbps_fmt_cb *cb, void *data, FILE *fp)
+{
+	struct strbuf buf = {0};
+	const char *pos = format;
+	int r = 0;
+
+	for (;;) {
+		struct xbps_fmt_spec spec;
+		r = nexttok(&pos, &buf);
+		if (r <= 0)
+			goto out;
+		switch (r) {
+		case TTEXT:
+			fprintf(fp, "%s", buf.mem);
+			break;
+		case TVAR:
+			r = parse(&pos, &buf, &spec);
+			if (r < 0)
+				goto out;
+			r = cb(fp, &spec, buf.mem, data);
+			if (r != 0)
+				goto out;
+			break;
+		}
+	}
+out:
+	strbuf_release(&buf);
+	return r;
+}
+
+int
+xbps_fmt(const struct xbps_fmt *fmt, xbps_fmt_cb *cb, void *data, FILE *fp)
+{
+	int r;
+	for (const struct xbps_fmt *f = fmt; f->common; f++) {
+		switch (f->common->type) {
+		case TTEXT:
+			fprintf(fp, "%s", f->chunk->s);
+			break;
+		case TVAR:
+			r = cb(fp, &f->var->spec, f->var->s, data);
+			if (r != 0)
+				return r;
+			break;
+		}
+	}
+	return 0;
+}
+
+struct fmt_dict_cb {
+	xbps_dictionary_t dict;
+};
+
+int
+xbps_fmt_string(const struct xbps_fmt_spec *spec, const char *str, size_t len, FILE *fp)
+{
+	if (len == 0)
+		len = strlen(str);
+	if (spec->align == '>' && spec->width > (unsigned int)len) {
+		for (unsigned i = 0; i < spec->width - len; i++)
+			fputc(spec->fill, fp);
+	}
+	fprintf(fp, "%s", str);
+	if (spec->align == '<' && spec->width > (unsigned int)len) {
+		for (unsigned i = 0; i < spec->width - len; i++)
+			fputc(spec->fill, fp);
+	}
+	return 0;
+}
+
+int
+xbps_fmt_number(const struct xbps_fmt_spec *spec, int64_t d, FILE *fp)
+{
+	char buf[20];
+	int len;
+
+	switch (spec->type) {
+	default: /* fallthrough */
+	case 'u': len = snprintf(buf, sizeof(buf), "%" PRIu64, d); break;
+	case 'd':
+		if (spec->sign == '+')
+			len = snprintf(buf, sizeof(buf), "%+" PRId64, d);
+		else
+			len = snprintf(buf, sizeof(buf), "%" PRId64, d);
+		if (spec->align == '=' && (buf[0] == '+' || buf[0] == '-')) {
+			struct xbps_fmt_spec s = *spec;
+			s.align = '>';
+			s.width -= 1;
+			fputc(buf[0], fp);
+			return xbps_fmt_string(&s, buf+1, len-1, fp);
+		}
+		break;
+	case 'x': len = snprintf(buf, sizeof(buf), "%" PRIx64, d); break;
+	case 'X': len = snprintf(buf, sizeof(buf), "%" PRIX64, d); break;
+	case 'h':
+		len = humanize_number(buf, spec->prec, d, "B",
```

Comment:
should #445 be considered here?

Reply:
Makes sense, switched `humanize_numbers(3)` to the freebsd version that supports it. We can switch the other places that use `xbps_humanize_number` later.